File: tests/experiments/tool-baseline/texlab_probe.py
#!/usr/bin/env python3
"""Collect texlab diagnostics for a LaTeX file.

texlab ships no analysis CLI — its only commands are `run` (the language server
over stdin/stdout) and `inverse-search`. To find out what it actually reports,
the server has to be driven over LSP. This script does that: initialize, open
the document, wait for `textDocument/publishDiagnostics`, shut down.

Usage:  python3 texlab_probe.py <file.tex> [seconds]
Output: one JSON object on stdout, {"file": …, "diagnostics": [...]}.
"""

import json
import os
import subprocess
import sys
import threading
import time


def frame(payload: dict) -> bytes:
    body = json.dumps(payload).encode("utf-8")
    return b"Content-Length: %d\r\n\r\n%s" % (len(body), body)


def read_messages(stream, sink, stop):
    """Read Content-Length framed JSON-RPC messages until the stream closes."""
    while not stop.is_set():
        header = b""
        while b"\r\n\r\n" not in header:
            byte = stream.read(1)
            if not byte:
                return
            header += byte
        length = 0
        for line in header.decode("utf-8", "replace").split("\r\n"):
            if line.lower().startswith("content-length:"):
                length = int(line.split(":", 1)[1].strip())
        body = stream.read(length)
        if not body:
            return
        try:
            sink.append(json.loads(body))
        except ValueError:
            pass


def main() -> int:
    if len(sys.argv) < 2:
        print(__doc__, file=sys.stderr)
        return 2
    path = os.path.abspath(sys.argv[1])
    settle = float(sys.argv[2]) if len(sys.argv) > 2 else 4.0
    root = os.path.dirname(path)

    proc = subprocess.Popen(
        ["texlab", "run"],
        stdin=subprocess.PIPE,
        stdout=subprocess.PIPE,
        stderr=subprocess.DEVNULL,
    )
    received: list = []
    stop = threading.Event()
    reader = threading.Thread(
        target=read_messages, args=(proc.stdout, received, stop), daemon=True
    )
    reader.start()

    def send(payload):
        proc.stdin.write(frame(payload))
        proc.stdin.flush()

    send({
        "jsonrpc": "2.0", "id": 1, "method": "initialize",
        "params": {
            "processId": os.getpid(),
            "rootUri": "file://" + root,
            "capabilities": {
                "textDocument": {"publishDiagnostics": {"relatedInformation": True}}
            },
        },
    })
    time.sleep(1.0)
    send({"jsonrpc": "2.0", "method": "initialized", "params": {}})

    with open(path, "r", encoding="utf-8", errors="replace") as handle:
        text = handle.read()
    send({
        "jsonrpc": "2.0", "method": "textDocument/didOpen",
        "params": {
            "textDocument": {
                "uri": "file://" + path,
                "languageId": "latex",
                "version": 1,
                "text": text,
            }
        },
    })

    time.sleep(settle)
    send({"jsonrpc": "2.0", "id": 2, "method": "shutdown", "params": None})
    time.sleep(0.4)
    send({"jsonrpc": "2.0", "method": "exit", "params": None})
    stop.set()
    try:
        proc.wait(timeout=3)
    except subprocess.TimeoutExpired:
        proc.kill()

    diagnostics = [
        d
        for message in received
        if message.get("method") == "textDocument/publishDiagnostics"
        for d in message.get("params", {}).get("diagnostics", [])
    ]
    print(json.dumps(
        {"file": os.path.basename(path), "diagnostics": diagnostics},
        ensure_ascii=False,
    ))
    return 0


if __name__ == "__main__":
    sys.exit(main())

File: tests/experiments/corpus-200/E3-spike-in/texlab-probe.mjs
// Copy of exacttex/repo/tests/experiments/0b-error-classes/texlab-probe.mjs, parametrized:
//   node texlab-probe.mjs <root-dir> <root.tex> [wait-ms]
// Opens the root file in texlab, collects every diagnostic published for any file, prints one per line
// as: <SEVERITY> <file> L<line>: <message>. English output; "(no diagnostics)" when none.
import { spawn } from "node:child_process";
import { readFileSync } from "node:fs";
import { resolve } from "node:path";

const dir = resolve(process.argv[2]);
const file = resolve(dir, process.argv[3] ?? "main.tex");
const waitMs = Number(process.argv[4] ?? 8000);
const server = spawn("texlab", [], { stdio: ["pipe", "pipe", "ignore"] });
const send = (msg) => {
  const body = JSON.stringify(msg);
  server.stdin.write(`Content-Length: ${Buffer.byteLength(body)}\r\n\r\n${body}`);
};
let buffer = Buffer.alloc(0);
const diagnostics = new Map(); // uri -> latest list
server.stdout.on("data", (chunk) => {
  buffer = Buffer.concat([buffer, chunk]);
  for (;;) {
    const head = buffer.indexOf("\r\n\r\n");
    if (head < 0) return;
    const length = Number(/Content-Length: (\d+)/.exec(buffer.slice(0, head))?.[1]);
    if (buffer.length < head + 4 + length) return;
    const msg = JSON.parse(buffer.slice(head + 4, head + 4 + length).toString());
    buffer = buffer.slice(head + 4 + length);
    if (msg.method === "textDocument/publishDiagnostics") {
      diagnostics.set(msg.params.uri, msg.params.diagnostics);
    }
  }
});
send({ jsonrpc: "2.0", id: 1, method: "initialize", params: { processId: null, rootUri: `file://${dir}`, capabilities: {} } });
setTimeout(() => {
  send({ jsonrpc: "2.0", method: "initialized", params: {} });
  send({ jsonrpc: "2.0", method: "textDocument/didOpen", params: { textDocument: { uri: `file://${file}`, languageId: "latex", version: 1, text: readFileSync(file, "utf8") } } });
}, 300);
setTimeout(() => {
  const lines = [];
  for (const [uri, ds] of diagnostics) {
    const rel = decodeURIComponent(uri.replace(`file://${dir}/`, ""));
    for (const d of ds) lines.push(`${d.severity === 1 ? "ERROR" : d.severity === 2 ? "WARN" : "info"} ${rel} L${d.range.start.line + 1}: ${d.message}`);
  }
  console.log(lines.length ? lines.join("\n") : "(no diagnostics)");
  server.kill();
  process.exit(0);
}, waitMs);
